README.md
# Copy Reference Advanced

Copy selected line number(s) with absolute/relative file path

result e.g. `path/to/file.txt:1-10`

## Hints

<!-- @watch package.json 55-60 -->
- Line numbers start with 1 by default (see Configuration to change this)
- if nothing is selected the line of the cursor is taken

## Features

Just what the description says: copy line numbers with absolute/relative file path

<!-- @watch package.json 34-37 -->
<!-- @watch package.json 44-47 -->
- Use `ctrl+shift+c` to copy the file path & selected line numbers  

<!-- @watch package.json 38-41 -->
<!-- @watch package.json 48-51 -->
- Use `ctrl+shift+a` to copy the file path & selected line numbers with the wrapping additional texts (see `copyReferenceAdvanced.prependToCopiedText` and `copyReferenceAdvanced.appendToCopiedText`)

or
<!-- @watch package.json 34-37 -->
use the command palette search for `CopyReference Advanced` or `CopyReference Advanced with additional text`

## Configuration

<!-- @watch package.json 56-60 -->
- `copyReferenceAdvanced.lineOffset` The line number to start with (this is added to the zero-based value of the line number)

<!-- @watch package.json 61-65 -->
- `copyReferenceAdvanced.pathLineNumberDelimiter` The delimiter for the path and line number section e.g. `-` would give `some/path.txt`**-**`1-5`, `:` would give `some/path.txt`**:**`1-5`

<!-- @watch package.json 66-70 -->
- `copyReferenceAdvanced.lineNumberDelimiter` The delimiter for the line numbers e.g. `--` would give `1--5`

<!-- @watch package.json 71-75 -->
- `copyReferenceAdvanced.collapseSingleLineNumbers` True: when only one line was selected e.g. 16 the output will be 16 without the line number delimiter, false: the output will always include the line number delimiter

<!-- @watch package.json 76-80 -->
- `copyReferenceAdvanced.prependToPath` Some string to prepend to the file path

<!-- @watch package.json 81-85 -->
- `copyReferenceAdvanced.appendToPath` Some string to append to the file path

<!-- @watch package.json 86-95 -->
- `copyReferenceAdvanced.pathMode` : `absolutePaths` will copy absolute file paths, `relativePaths` will copy relative file paths if the file is inside a opened workspace folder, `relativePathsWithoutFolderName` the same as `relativePaths` but the folder name is removed

<!-- @watch package.json 96-100 -->
- `copyReferenceAdvanced.failIfNoFolderIsOpen` If `pathMode` is set to relative and no folder is open... True: notify warning because we cannot get a relative path (noting is copied), false: just copy the absolute path

<!-- @watch package.json 101-105 -->
- `copyReferenceAdvanced.displayCopiedTimeoutInMs` The copied text is displayed as a status bar item for X ms. Use a value <= 0 to always hide the status bar item

<!-- @watch package.json 106-110 -->
- `copyReferenceAdvanced.prependToCopiedText` Some text that should be prepended to the whole copied text

<!-- @watch package.json 111-115 -->
- `copyReferenceAdvanced.appendToCopiedText` Some text that should be appended to the whole copied text

## Known Issues

## Release Notes

See [Changelog](CHANGELOG.md)

## Install extension locally

https://code.visualstudio.com/Docs/editor/extension-gallery#_common-questions

macOS: `~/.vscode/extensions`


CHANGELOG.md
# Change Log

<!-- @watch package.json 5 -->

## 1.0.0

- Initial release

src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { env, workspace, window, WorkspaceFolder, StatusBarItem, StatusBarAlignment } from 'vscode';
import * as ncp from "copy-paste";
import { setTimeout } from 'timers';


type PathMode = "absolutePaths" | "relativePaths" | "relativePathsWithoutFolderName"

let appDisplayName = "[CopyReference Advanced]"

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('CopyReference Advanced is now active!');

    let statusBarHelper = new StatusBarHelper()

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json

    /* @watch package.json 34-37 */
    /* @watch package.json 44-47 */
    let disposable = vscode.commands.registerCommand('extension.copyReferenceAdvanced', () => {
        // The code you place here will be executed every time your command is executed
        copyReferenceAdvanced(statusBarHelper, false, null)

    });

    /* @watch package.json 38-41 */
    /* @watch package.json 48-51 */
    let disposable2 = vscode.commands.registerCommand('extension.copyReferenceAdvancedWithAdditionalText', () => {
        // The code you place here will be executed every time your command is executed
        copyReferenceAdvanced(statusBarHelper, true, null)
    });

    let disposable3 = vscode.commands.registerCommand('extension.copyReferenceAdvancedIgnoreLineNumbers', () => {
        // The code you place here will be executed every time your command is executed
        copyReferenceAdvanced(statusBarHelper, false, false)
    });

    let disposable4 = vscode.commands.registerCommand('extension.copyReferenceAdvancedWithAdditionalTextIgnoreLineNumbers', () => {
        // The code you place here will be executed every time your command is executed
        copyReferenceAdvanced(statusBarHelper, true, false)
    });

    context.subscriptions.push(disposable);
    context.subscriptions.push(disposable2);
    context.subscriptions.push(disposable3);
    context.subscriptions.push(disposable4);
    context.subscriptions.push(statusBarHelper);
}

/**
 * main method
 * @param statusBarHelper the status bar helper
 * @param useAppendTexts true: use the append to copied texts if any, false: not (may be used for another shortcut...)
 * @param copyLineNumbersToo true: force copy line numbers (ignore settings), false: force don't copy line numbers (ignore copy), null: use setting
 */
function copyReferenceAdvanced(statusBarHelper: StatusBarHelper, useAppendTexts: boolean, copyLineNumbersToo: boolean | null) {

    let editor = vscode.window.activeTextEditor

    if (!editor) {
        vscode.window.showWarningMessage(`${appDisplayName} 1. Open a file (in a folder) and execute the command`)
        return
    }

    //get settings before every call else we might get outdated values
    //(if the user changed settings and didn't restart vs code)

    let config = vscode.workspace.getConfiguration('copyReferenceAdvanced')

    /* @watch package.json 56-60 */
    let lineOffset: number = config.get('lineOffset')
    /* @watch package.json 61-65 */
    let pathLineNumberDelimiter: string = config.get('pathLineNumberDelimiter')
    /* @watch package.json 66-70 */
    let lineNumberDelimiter: string = config.get('lineNumberDelimiter')
    /* @watch package.json 71-75 */
    let collapseSingleLineNumbers: boolean = config.get('collapseSingleLineNumbers')
    /* @watch package.json 76-80 */
    let prependToPath: string = config.get('prependToPath')
    /* @watch package.json 81-85 */
    let appendToPath: string = config.get('appendToPath')
    /* @watch package.json 86-95 */
    let pathMode: PathMode = config.get('pathMode')
    /* @watch package.json 96-100 */
    let failIfNoFolderIsOpen: boolean = config.get('failIfNoFolderIsOpen')
    /* @watch package.json 101-105 */
    let displayCopiedTimeoutInMs: number = config.get('displayCopiedTimeoutInMs')
    /* @watch package.json 106-110 */
    let prependToCopiedText: string = config.get('prependToCopiedText')
    /* @watch package.json 111-115 */
    let appendToCopiedText: string = config.get('appendToCopiedText')

    let prependToLines: string = config.get('prependToLines')

    let appendToLines: string = config.get('appendToLines')

    let copyLineNumbers: boolean = config.get('copyLineNumbers')

    if (copyLineNumbersToo !== null) {
        copyLineNumbers = copyLineNumbersToo
    }


    let selection = editor.selection

    let startLine = selection.start.line + lineOffset;
    let endLine = selection.end.line + lineOffset;

    let linesRange = ''

    linesRange = selection.isSingleLine && collapseSingleLineNumbers
        ? `${startLine}`
        : `${startLine}${lineNumberDelimiter}${endLine}`

    linesRange = `${prependToLines}${linesRange}${appendToLines}`

    let fileAbsolutePath = editor.document.fileName

    if (useAppendTexts === false) {
        prependToCopiedText = ""
        appendToCopiedText = ""
    }

    let resultString = "error occured"

    let rootPathAbsolute: WorkspaceFolder = null
    rootPathAbsolute = workspace.getWorkspaceFolder(editor.document.uri)

    if (pathMode === "absolutePaths") {
        let pathPart = `${prependToPath}${fileAbsolutePath}${appendToPath}`

        resultString = copyLineNumbers
            ? `${prependToCopiedText}${pathPart}${pathLineNumberDelimiter}${linesRange}${appendToCopiedText}`
            : `${prependToCopiedText}${pathPart}${appendToCopiedText}`

        copyToClipboard(resultString, statusBarHelper, displayCopiedTimeoutInMs)
        return
    }

    if (rootPathAbsolute === undefined) {
        //no folder is open 

        if (failIfNoFolderIsOpen) {
            vscode.window.showWarningMessage(`${appDisplayName} You must open a folder for this action to work (because you specified relative paths and failIfNoFolderIsOpen is set to true).`)
            return
        }

        let pathPart = `${prependToPath}${fileAbsolutePath}${appendToPath}`

        resultString = copyLineNumbers
            ? `${prependToCopiedText}${pathPart}${pathLineNumberDelimiter}${linesRange}${appendToCopiedText}`
            : `${prependToCopiedText}${pathPart}${appendToCopiedText}`

        copyToClipboard(resultString, statusBarHelper, displayCopiedTimeoutInMs)
        return
    }

    //--- we need to go relative 



    if (pathMode === "relativePaths") {
        let relativePath = workspace.asRelativePath(fileAbsolutePath, true)

        let pathPart = `${prependToPath}${relativePath}${appendToPath}`

        resultString = copyLineNumbers
            ? `${prependToCopiedText}${pathPart}${pathLineNumberDelimiter}${linesRange}${appendToCopiedText}`
            : `${prependToCopiedText}${pathPart}${appendToCopiedText}`

        copyToClipboard(resultString, statusBarHelper, displayCopiedTimeoutInMs)
        return
    }
    else if (pathMode === "relativePathsWithoutFolderName") {
        let relativePath = workspace.asRelativePath(fileAbsolutePath, false)

        let pathPart = `${prependToPath}${relativePath}${appendToPath}`

        resultString = copyLineNumbers
            ? `${prependToCopiedText}${pathPart}${pathLineNumberDelimiter}${linesRange}${appendToCopiedText}`
            : `${prependToCopiedText}${pathPart}${appendToCopiedText}`

        copyToClipboard(resultString, statusBarHelper, displayCopiedTimeoutInMs)
        return
    }
    else {
        vscode.window.showWarningMessage(`${appDisplayName} Invalid option for copyReferenceAdvanced.pathMode specified!`)
        return
    }

}

function copyToClipboard(text: string, statusBarHelper: StatusBarHelper, timeoutInMs: number) {
    ncp.copy(text, () => {
        //console.log('copied!')
        statusBarHelper.updateStatusBarItem(`Copied reference: ${text}`, timeoutInMs)
    })
}

class StatusBarHelper {
    private _statusBarItem: StatusBarItem;

    //https://code.visualstudio.com/docs/extensions/example-word-count#_update-the-status-bar

    /**
     * 
     * @param text displays the status bar item
     * @param timeoutInS hide after x seconds, <= 0 to not show status bar item
     */
    public updateStatusBarItem(text: string, timeoutInMs: number) {
        // Create as needed
        if (!this._statusBarItem) {
            this._statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);
        }

        if (timeoutInMs <= 0) {
            this._statusBarItem.hide();
            return;
        }

        // Get the current text editor
        let editor = window.activeTextEditor;
        if (!editor) {
            this._statusBarItem.hide();
            return;
        }

        // Update the status bar
        this._statusBarItem.text = text;
        this._statusBarItem.show();

        setTimeout(() => {
            this._statusBarItem.hide();
        }, timeoutInMs)

    }

    dispose() {
        this._statusBarItem.dispose();
    }
}

// this method is called when your extension is deactivated
export function deactivate() {
    console.log(`CopyReference Advanced is now deactivated!`);
}